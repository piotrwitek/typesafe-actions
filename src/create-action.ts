import {
  TypeConstant,
  ActionCreatorBuilder,
  ActionCreatorTypeMetadata,
  ActionBuilder,
} from './type-helpers';
import { createCustomAction } from './create-custom-action';
import {
  checkIsEmpty,
  throwIsEmpty,
  checkInvalidActionType,
  throwInvalidActionType,
} from './utils/validation';

export function createAction<TType extends TypeConstant>(
  type: TType
): <TPayload = undefined, TMeta = undefined>() => ActionCreatorBuilder<
  TType,
  TPayload,
  TMeta
>;

export function createAction<
  TType extends TypeConstant,
  TCreatorPayload extends any = undefined,
  TCreatorMeta extends any = undefined,
  TArgs extends any[] = any[]
>(
  type: TType,
  payloadCreator: undefined | ((...args: TArgs) => TCreatorPayload),
  metaCreator?: (...args: TArgs) => TCreatorMeta
): <
  TPayload extends TCreatorPayload = TCreatorPayload,
  TMeta extends TCreatorMeta = TCreatorMeta
>() => (...args: TArgs) => ActionBuilder<TType, TPayload, TMeta>;

/**
 * @description create an action-creator
 */
export function createAction<
  TType extends TypeConstant,
  TCreatorPayload extends any = undefined,
  TCreatorMeta extends any = undefined,
  TArgs extends any[] = any[]
>(
  type: TType,
  payloadCreator?: undefined | ((...args: TArgs) => TCreatorPayload),
  metaCreator?: (...args: TArgs) => TCreatorMeta
):
  | (<
      TPayload extends TCreatorPayload = TCreatorPayload,
      TMeta extends TCreatorMeta = TCreatorMeta
    >() => (...args: TArgs) => ActionBuilder<TType, TPayload, TMeta>)
  | (<TPayload = undefined, TMeta = undefined>() => ActionCreatorBuilder<
      TType,
      TPayload,
      TMeta
    >) {
  if (checkIsEmpty(type)) {
    throwIsEmpty(1);
  }

  if (checkInvalidActionType(type)) {
    throwInvalidActionType(1);
  }

  return <TPayload, TMeta = undefined>() => {
    return createCustomAction(type, (...args: TArgs) => {
      let payload = args[0];
      let meta = args[1];

      if (payloadCreator != null || metaCreator != null) {
        payload = payloadCreator != null ? payloadCreator(...args) : undefined;
        meta = metaCreator != null ? metaCreator(...args) : undefined;
      }

      return {
        ...(payload !== undefined && { payload }),
        ...(meta !== undefined && { meta }),
      };
    }) as ActionCreatorBuilder<TType, TPayload, TMeta> &
      ActionCreatorTypeMetadata<
        TType,
        ReturnType<ActionCreatorBuilder<TType, TPayload, TMeta>>
      >;
  };
}
