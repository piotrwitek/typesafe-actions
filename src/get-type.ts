import {
  TypeConstant,
  ActionCreator,
  ActionCreatorTypeMetadata,
} from './type-helpers';
import {
  checkIsEmpty,
  throwIsEmpty,
  checkInvalidActionCreator,
  throwInvalidActionCreator,
} from './utils/validation';

/**
 * @description get the "type literal" of a given action-creator
 */
export function getType<TType extends TypeConstant>(
  actionCreator: ActionCreator<TType> & ActionCreatorTypeMetadata<TType>
): TType {
  if (checkIsEmpty(actionCreator)) {
    throwIsEmpty(1);
  }

  if (checkInvalidActionCreator(actionCreator)) {
    throwInvalidActionCreator(1);
  }

  return actionCreator.getType();
}
