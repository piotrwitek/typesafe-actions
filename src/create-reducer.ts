import { getType } from './get-type';
import {
  checkValidActionCreator,
  checkValidActionType,
  throwInvalidActionTypeOrActionCreator,
} from './utils/validation';
import { Reducer, Action, Types } from './type-helpers';

export type HandleActionChainApi<
  TState,
  TInputAction extends Action,
  TRootAction extends Action
> = <
  TActionCreator extends (...args: any[]) => TInputAction,
  THandledAction extends ReturnType<TActionCreator>,
  TOutputAction extends Exclude<TInputAction, THandledAction>
>(
  singleOrMultipleCreatorsAndTypes: TActionCreator | TActionCreator[],
  reducer: (state: TState, action: THandledAction) => TState
) => [TOutputAction] extends [Action]
  ? Reducer<TState, TRootAction> & {
      handlers: Record<
        Exclude<TRootAction, TOutputAction>['type'],
        (state: TState, action: TRootAction) => TState
      >;
      handleAction: HandleActionChainApi<TState, TOutputAction, TRootAction>;
      defaultHandler: HandleDefaultActionChainApi<
        TState,
        TOutputAction,
        TRootAction
      >;
    }
  : Reducer<TState, TRootAction> & {
      handlers: Record<
        TRootAction['type'],
        (state: TState, action: TRootAction) => TState
      >;
    };

type HandleDefaultActionChainApi<
  TState,
  TInputAction extends Action,
  TRootAction extends Action
> = <
  TActionCreator extends (...args: any[]) => TInputAction,
  THandledAction extends ReturnType<TActionCreator>,
  TOutputAction extends Exclude<TInputAction, THandledAction>
>(
  reducer: (state: TState, action: THandledAction) => TState
) => [TOutputAction] extends [Action]
  ? Reducer<TState, TRootAction> & {
      handlers: Record<
        Exclude<TRootAction, TOutputAction>['type'],
        (state: TState, action: TRootAction) => TState
      >;
    }
  : Reducer<TState, TRootAction> & {
      handlers: Record<
        TRootAction['type'],
        (state: TState, action: TRootAction) => TState
      >;
    };

export type HandleTypeChainApi<
  TState,
  TInputAction extends Action,
  TRootAction extends Action
> = <
  TType extends TInputAction['type'],
  THandledAction extends Extract<TInputAction, Action<TType>>,
  TOutputAction extends Exclude<TInputAction, THandledAction>
>(
  singleOrMultipleCreatorsAndTypes: TType | TType[],
  reducer: (state: TState, action: THandledAction) => TState
) => [TOutputAction] extends [Action]
  ? Reducer<TState, TRootAction> & {
      handlers: Record<
        Exclude<TRootAction, TOutputAction>['type'],
        (state: TState, action: TRootAction) => TState
      >;
      handleType: HandleTypeChainApi<TState, TOutputAction, TRootAction>;
      defaultHandler: HandleDefaultActionChainApi<
        TState,
        TOutputAction,
        TRootAction
      >;
    }
  : Reducer<TState, TRootAction> & {
      handlers: Record<
        TRootAction['type'],
        (state: TState, action: TRootAction) => TState
      >;
    };

type GetAction<
  TAction extends Action,
  TType extends TAction['type']
> = TAction extends Action<TType> ? TAction : never;

export type InitialHandler<TState, TRootAction extends Action> = {
  [P in TRootAction['type']]?: (
    state: TState,
    action: GetAction<TRootAction, P>
  ) => TState;
};

type RootAction = Types extends { RootAction: infer T } ? T : any;

export function createReducer<TState, TRootAction extends Action = RootAction>(
  initialState: TState,
  initialHandlers: InitialHandler<TState, TRootAction> = {},
  defaultReducer?: Reducer<TState, TRootAction>
) {
  const handlers: any = {
    ...initialHandlers,
  };

  const rootReducer: Reducer<TState, TRootAction> = (
    state = initialState,
    action: TRootAction
  ) => {
    if (handlers.hasOwnProperty(action.type)) {
      const reducer = handlers[action.type];
      if (typeof reducer !== 'function') {
        throw Error(
          `Reducer under "${action.type}" key is not a valid reducer`
        );
      }
      return reducer(state, action);
    } else if (defaultReducer) {
      return defaultReducer(state, action);
    } else {
      return state;
    }
  };

  const reducerHandler = ((
    singleOrMultipleCreatorsAndTypes: any,
    reducer: any
  ) => {
    const creatorsAndTypes = Array.isArray(singleOrMultipleCreatorsAndTypes)
      ? singleOrMultipleCreatorsAndTypes
      : [singleOrMultipleCreatorsAndTypes];

    const newHandlers: typeof handlers = {};

    creatorsAndTypes
      .map(
        (
          acOrType: TRootAction['type'] | ((...args: any[]) => TRootAction),
          index
        ) =>
          checkValidActionCreator(acOrType)
            ? getType(acOrType)
            : checkValidActionType(acOrType)
            ? acOrType
            : throwInvalidActionTypeOrActionCreator(index + 1)
      )
      .forEach(type => (newHandlers[type] = reducer));

    return createReducer<TState, TRootAction>(
      initialState,
      {
        ...handlers,
        ...newHandlers,
      },
      defaultReducer
    );
  }) as
    | HandleActionChainApi<TState, TRootAction, TRootAction>
    | HandleTypeChainApi<TState, TRootAction, TRootAction>;

  const defaultHandler = ((reducer: any) => {
    return createReducer<TState, TRootAction>(initialState, handlers, reducer);
  }) as HandleDefaultActionChainApi<TState, TRootAction, TRootAction>;

  const chainApi = Object.assign(rootReducer, {
    handlers: { ...handlers },
    handleAction: reducerHandler,
    handleType: reducerHandler,
    defaultHandler,
  }) as Reducer<TState, TRootAction> &
    Readonly<{
      handlers: InitialHandler<TState, RootAction>;
      handleAction: [unknown] extends [TRootAction]
        ? any
        : HandleActionChainApi<TState, TRootAction, TRootAction>;
      handleType: [unknown] extends [TRootAction]
        ? any
        : HandleTypeChainApi<TState, TRootAction, TRootAction>;
      defaultHandler: [unknown] extends [TRootAction]
        ? any
        : HandleDefaultActionChainApi<TState, TRootAction, TRootAction>;
    }>;

  return chainApi;
}
