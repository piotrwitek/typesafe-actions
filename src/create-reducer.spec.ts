import * as T from './type-helpers'; // type-tests global
import { ActionType } from './type-helpers';
import { createAction } from './create-action';
import { createReducer } from './create-reducer';
import { getType } from './get-type';

const add = createAction('ADD')<number>();
const increment = createAction('INCREMENT')();
const decrement = createAction('DECREMENT')();
const reduxInit = createAction('@@redux/INIT.1')();
const actions = {
  add,
  increment,
  decrement,
  reduxInit,
};

declare module './type-helpers' {
  interface Types {
    RootAction: ActionType<typeof actions>;
  }
}

const initialState = 0;

// @dts-jest:group With Action Creators
{
  const emptyReducer = createReducer(initialState);

  const counterReducer1 = emptyReducer.handleAction(
    [add, increment],
    (state, action) => state + (action.type === 'ADD' ? action.payload : 1)
  );
  // @dts-jest:pass:snap
  counterReducer1.handlers;
  // @dts-jest:pass
  Object.keys({ ...counterReducer1.handlers }); // => ["ADD", "INCREMENT"]
  // @dts-jest:pass
  Object.keys({ ...emptyReducer.handlers }); // => []

  const counterReducer2 = emptyReducer
    .handleAction([add], (state, action) => state + action.payload)
    .handleAction([increment], (state, _) => state + 1);
  // @dts-jest:pass:snap
  counterReducer2.handlers;
  // @dts-jest:pass
  Object.keys({ ...counterReducer2.handlers }); // => ["ADD", "INCREMENT"]
  // @dts-jest:pass
  Object.keys({ ...emptyReducer.handlers }); // => []

  const counterReducer3 = emptyReducer.handleAction(
    add,
    (state, action) => state + action.payload
  );
  // @dts-jest:pass:snap
  counterReducer3.handlers;
  // @dts-jest:pass
  Object.keys({ ...counterReducer3.handlers }); // => [ "ADD"]
  // @dts-jest:pass
  Object.keys({ ...emptyReducer.handlers }); // => []

  const counterReducer4 = emptyReducer.handleAction(
    increment,
    (state, _) => state + 1
  );
  // @dts-jest:pass:snap
  counterReducer4.handlers;
  // @dts-jest:pass
  Object.keys({ ...counterReducer4.handlers }); // => [ "INCREMENT"]
  // @dts-jest:pass
  Object.keys({ ...emptyReducer.handlers }); // => []

  const counterReducer5 = emptyReducer
    .handleAction(add, (state, action) => state + action.payload)
    .defaultHandler((state, action) => state + 1);

  // @dts-jest:pass:snap
  counterReducer5.handlers;
  // @dts-jest:pass
  Object.keys({ ...counterReducer5.handlers }); // => [ "ADD"]
  // @dts-jest:pass
  Object.keys({ ...emptyReducer.handlers }); // => []

  {
    [
      counterReducer1,
      counterReducer2,
      createReducer(initialState, {
        [getType(add)]: counterReducer3.handlers.ADD,
        [getType(increment)]: counterReducer4.handlers.INCREMENT,
      }),
      // has to be typed as any because of typescript bug: https://github.com/microsoft/TypeScript/issues/29718
      createReducer(initialState, {
        [getType(add)]: (state: any, action: any) => state + action.payload,
        [getType(increment)]: (state: any, _: any) => state + 1,
      }),
      createReducer<number, ActionType<typeof actions>>(initialState, {
        [getType(add)]: (state: any, action: any) => state + action.payload,
        [getType(increment)]: (state: any, _: any) => state + 1,
      }),
    ].forEach(fn => {
      // @dts-jest:pass
      fn(0, {} as any); // => 0
      // @dts-jest:pass
      fn(0, increment()); // => 1
      // @dts-jest:pass
      fn(0, add(4)); // => 4
    });
    // @dts-jest:pass
    counterReducer5(0, {} as any); // => 1
    // @dts-jest:pass
    counterReducer5(0, add(4)); // => 4
  }
}

// @dts-jest:group Type refinement checks
{
  type State = {
    foo: string | null;
  };

  const defaultState: State = {
    foo: null,
  };

  const actions2 = {
    foo1: createAction('foo1')<string>(),
    foo2: createAction('foo2')<string>(),
    foo3: createAction('foo3')(),
    foo4: createAction('foo4')(),
    foo5: createAction('foo5')(),
  };

  type Action = ActionType<typeof actions2>;

  const reducer = createReducer<State, Action>(defaultState)
    .handleAction(actions2.foo1, (state, action) => {
      // @dts-jest:pass:snap
      state;
      // @dts-jest:pass:snap
      action;

      return {
        ...state,
        foo: action.payload,
      };
    })
    .handleAction(actions2.foo2, (state, action) => {
      // @dts-jest:pass:snap
      state;
      // @dts-jest:pass:snap
      action;

      return {
        ...state,
        foo: action.payload,
      };
    })
    .handleAction(actions2.foo3, (state, action) => {
      // @dts-jest:pass:snap
      state;
      // @dts-jest:pass:snap
      action;

      return {
        ...state,
        foo: 'empty',
      };
    })
    .handleAction(actions2.foo4, (state, action) => {
      // @dts-jest:pass:snap
      state;
      // @dts-jest:pass:snap
      action;

      return {
        ...state,
        foo: 'empty',
      };
    })
    .defaultHandler((state, action) => {
      // @dts-jest:pass:snap
      state;
      // @dts-jest:pass:snap
      action;

      return {
        ...state,
        foo: 'default',
      };
    });

  // @dts-jest:pass:snap
  reducer.handlers;

  {
    [
      reducer(defaultState, actions2.foo1('check')),
      reducer(defaultState, actions2.foo2('check')),
    ].forEach(reducerResult => {
      // @dts-jest:pass:snap
      reducerResult; // => { foo: "check" }
    });

    [
      reducer(defaultState, actions2.foo3()),
      reducer(defaultState, actions2.foo4()),
    ].forEach(reducerResult => {
      // @dts-jest:pass:snap
      reducerResult; // => { foo: "empty" }
    });

    const reducerDefaultResult = reducer(defaultState, actions2.foo5());
    // @dts-jest:pass:snap
    reducerDefaultResult; // => { foo: "default" }
  }
}

// @dts-jest:group With Action Types
{
  const reducerTest = createReducer(initialState);

  const counterReducer1 = reducerTest.handleType(
    ['ADD', 'INCREMENT'],
    (state, action) => state + (action.type === 'ADD' ? action.payload : 1)
  );
  // @dts-jest:pass:snap
  counterReducer1.handlers;
  // @dts-jest:pass
  Object.keys({ ...counterReducer1.handlers }); // => ["ADD", "INCREMENT"]
  // @dts-jest:pass
  Object.keys({ ...reducerTest.handlers }); // => []

  const counterReducer2 = reducerTest
    .handleType(['ADD'], (state, action) => state + action.payload)
    .handleType(['INCREMENT'], (state, _) => state + 1);
  // @dts-jest:pass:snap
  counterReducer2.handlers;
  // @dts-jest:pass
  Object.keys({ ...counterReducer2.handlers }); // => ["ADD", "INCREMENT"]
  // @dts-jest:pass
  Object.keys({ ...reducerTest.handlers }); // => []

  const counterReducer3 = reducerTest.handleType(
    'ADD',
    (state, action) => state + action.payload
  );
  // @dts-jest:pass:snap
  counterReducer3.handlers;
  // @dts-jest:pass
  Object.keys({ ...counterReducer3.handlers }); // => ["ADD"]
  // @dts-jest:pass
  Object.keys({ ...reducerTest.handlers }); // => []

  const counterReducer4 = reducerTest.handleType(
    'INCREMENT',
    (state, _) => state + 1
  );
  // @dts-jest:pass:snap
  counterReducer4.handlers;
  // @dts-jest:pass
  Object.keys({ ...counterReducer4.handlers }); // => [ "INCREMENT"]
  // @dts-jest:pass
  Object.keys({ ...reducerTest.handlers }); // => []

  const counterReducer5 = reducerTest
    .handleType(['ADD'], (state, action) => state + action.payload)
    .defaultHandler((state, action) => state + 1);
  // @dts-jest:pass:snap
  counterReducer5.handlers;
  // @dts-jest:pass
  Object.keys({ ...counterReducer5.handlers }); // => [ "ADD"]
  // @dts-jest:pass
  Object.keys({ ...reducerTest.handlers }); // => []

  {
    [
      counterReducer1,
      counterReducer2,
      createReducer(initialState, {
        ADD: counterReducer3.handlers.ADD,
        INCREMENT: counterReducer4.handlers.INCREMENT,
      }),
      createReducer(initialState, {
        ADD: (state, action) => state + action.payload,
        INCREMENT: (state, action) => state + 1,
      }),
      createReducer<number, ActionType<typeof actions>>(initialState, {
        ADD: (state, action) => state + action.payload,
        INCREMENT: (state, action) => state + 1,
      }),
    ].forEach(fn => {
      // @dts-jest:pass
      fn(0, {} as any); // => 0
      // @dts-jest:pass
      fn(0, increment()); // => 1
      // @dts-jest:pass
      fn(0, add(4)); // => 4
    });
    // @dts-jest:pass
    counterReducer5(0, increment()); // => 1;
    // @dts-jest:pass
    counterReducer5(0, add(4)); // => 4;
  }
}
