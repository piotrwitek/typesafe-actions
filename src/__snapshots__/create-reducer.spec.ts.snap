// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Type refinement checks action (type) should match snapshot 1`] = `"T.PayloadAction<\\"foo1\\", string>"`;

exports[`Type refinement checks action (type) should match snapshot 2`] = `"T.PayloadAction<\\"foo2\\", string>"`;

exports[`Type refinement checks action (type) should match snapshot 3`] = `"T.EmptyAction<\\"foo3\\">"`;

exports[`Type refinement checks action (type) should match snapshot 4`] = `"T.EmptyAction<\\"foo4\\">"`;

exports[`Type refinement checks action (type) should match snapshot 5`] = `"T.EmptyAction<\\"foo5\\">"`;

exports[`Type refinement checks reducer.handlers (type) should match snapshot 1`] = `"Record<\\"foo1\\" | \\"foo2\\" | \\"foo3\\" | \\"foo4\\" | \\"foo5\\", (state: { foo: string | null; }, action: T.PayloadAction<\\"foo1\\", string> | T.PayloadAction<\\"foo2\\", string> | T.EmptyAction<\\"foo3\\"> | T.EmptyAction<\\"foo4\\"> | T.EmptyAction<\\"foo5\\">) => { foo: string | null; }>"`;

exports[`Type refinement checks reducerDefaultResult (type) should match snapshot 1`] = `"{ foo: string | null; }"`;

exports[`Type refinement checks reducerResult (type) should match snapshot 1`] = `"{ foo: string | null; }"`;

exports[`Type refinement checks reducerResult (type) should match snapshot 2`] = `"{ foo: string | null; }"`;

exports[`Type refinement checks reducerResult (type) should match snapshot 3`] = `"{ foo: string | null; }"`;

exports[`Type refinement checks reducerResult (type) should match snapshot 4`] = `"{ foo: string | null; }"`;

exports[`Type refinement checks state (type) should match snapshot 1`] = `"{ foo: string | null; }"`;

exports[`Type refinement checks state (type) should match snapshot 2`] = `"{ foo: string | null; }"`;

exports[`Type refinement checks state (type) should match snapshot 3`] = `"{ foo: string | null; }"`;

exports[`Type refinement checks state (type) should match snapshot 4`] = `"{ foo: string | null; }"`;

exports[`Type refinement checks state (type) should match snapshot 5`] = `"{ foo: string | null; }"`;

exports[`With Action Creators counterReducer1.handlers (type) should match snapshot 1`] = `"Record<\\"ADD\\" | \\"INCREMENT\\", (state: number, action: T.PayloadAction<\\"ADD\\", number> | T.EmptyAction<\\"INCREMENT\\"> | T.EmptyAction<\\"DECREMENT\\"> | T.EmptyAction<\\"@@redux/INIT.1\\">) => number>"`;

exports[`With Action Creators counterReducer2.handlers (type) should match snapshot 1`] = `"Record<\\"ADD\\" | \\"INCREMENT\\", (state: number, action: T.PayloadAction<\\"ADD\\", number> | T.EmptyAction<\\"INCREMENT\\"> | T.EmptyAction<\\"DECREMENT\\"> | T.EmptyAction<\\"@@redux/INIT.1\\">) => number>"`;

exports[`With Action Creators counterReducer3.handlers (type) should match snapshot 1`] = `"Record<\\"ADD\\", (state: number, action: T.PayloadAction<\\"ADD\\", number> | T.EmptyAction<\\"INCREMENT\\"> | T.EmptyAction<\\"DECREMENT\\"> | T.EmptyAction<\\"@@redux/INIT.1\\">) => number>"`;

exports[`With Action Creators counterReducer4.handlers (type) should match snapshot 1`] = `"Record<\\"INCREMENT\\", (state: number, action: T.PayloadAction<\\"ADD\\", number> | T.EmptyAction<\\"INCREMENT\\"> | T.EmptyAction<\\"DECREMENT\\"> | T.EmptyAction<\\"@@redux/INIT.1\\">) => number>"`;

exports[`With Action Creators counterReducer5.handlers (type) should match snapshot 1`] = `"Record<\\"ADD\\" | \\"INCREMENT\\" | \\"DECREMENT\\" | \\"@@redux/INIT.1\\", (state: number, action: T.PayloadAction<\\"ADD\\", number> | T.EmptyAction<\\"INCREMENT\\"> | T.EmptyAction<\\"DECREMENT\\"> | T.EmptyAction<\\"@@redux/INIT.1\\">) => number>"`;

exports[`With Action Types counterReducer1.handlers (type) should match snapshot 1`] = `"Record<\\"ADD\\" | \\"INCREMENT\\", (state: number, action: T.PayloadAction<\\"ADD\\", number> | T.EmptyAction<\\"INCREMENT\\"> | T.EmptyAction<\\"DECREMENT\\"> | T.EmptyAction<\\"@@redux/INIT.1\\">) => number>"`;

exports[`With Action Types counterReducer2.handlers (type) should match snapshot 1`] = `"Record<\\"ADD\\" | \\"INCREMENT\\", (state: number, action: T.PayloadAction<\\"ADD\\", number> | T.EmptyAction<\\"INCREMENT\\"> | T.EmptyAction<\\"DECREMENT\\"> | T.EmptyAction<\\"@@redux/INIT.1\\">) => number>"`;

exports[`With Action Types counterReducer3.handlers (type) should match snapshot 1`] = `"Record<\\"ADD\\", (state: number, action: T.PayloadAction<\\"ADD\\", number> | T.EmptyAction<\\"INCREMENT\\"> | T.EmptyAction<\\"DECREMENT\\"> | T.EmptyAction<\\"@@redux/INIT.1\\">) => number>"`;

exports[`With Action Types counterReducer4.handlers (type) should match snapshot 1`] = `"Record<\\"INCREMENT\\", (state: number, action: T.PayloadAction<\\"ADD\\", number> | T.EmptyAction<\\"INCREMENT\\"> | T.EmptyAction<\\"DECREMENT\\"> | T.EmptyAction<\\"@@redux/INIT.1\\">) => number>"`;

exports[`With Action Types counterReducer5.handlers (type) should match snapshot 1`] = `"Record<\\"ADD\\" | \\"INCREMENT\\" | \\"DECREMENT\\" | \\"@@redux/INIT.1\\", (state: number, action: T.PayloadAction<\\"ADD\\", number> | T.EmptyAction<\\"INCREMENT\\"> | T.EmptyAction<\\"DECREMENT\\"> | T.EmptyAction<\\"@@redux/INIT.1\\">) => number>"`;
